### ISSUE TITLE
OSC problem...

### ISSUE THREAD

**Issue by rorywalsh** [csound/csound#704]
Some of Oeyvind's Cabbage instruments that make use of OSC are causing crashes in Csound when we try to destroy and recreate a Csound object. For example each time we save or update an instrument we recreate our Csound object. Throw OSC into the mix and strange things happen. The backtrace is below. The offending .csd file is a little complex, hopefully it's an easy fix. If not we can try to prepare a simplified example. 

0  0x00007ffff6c58d84  __GI___pthread_mutex_lock  /build/glibc-GKVZIf/glibc-2.23/nptl/../nptl/pthread_mutex_lock.c  67
1  0x00007ffff66b1349  csoundLockMutex  /home/rory/sourcecode/cabbageaudio/csound/Top/threads.c  496
2  0x00007fffe44f9c69  OSC_listdeinit    
3  0x00007ffff66a976e  csoundDeinitialiseOpcodes  /home/rory/sourcecode/cabbageaudio/csound/Top/csound.c  3762
4  0x00007ffff652505e  deact  /home/rory/sourcecode/cabbageaudio/csound/Engine/insert.c  727
5  0x00007ffff6525640  xturnoff  /home/rory/sourcecode/cabbageaudio/csound/Engine/insert.c  829
6  0x00007ffff6525695  xturnoff_now  /home/rory/sourcecode/cabbageaudio/csound/Engine/insert.c  842
7  0x00007ffff65314e4  deactivate_all_notes  /home/rory/sourcecode/cabbageaudio/csound/Engine/musmon.c  378
8  0x00007ffff65316d2  csoundCleanup  /home/rory/sourcecode/cabbageaudio/csound/Engine/musmon.c  432
9  0x00007ffff66a7b87  reset  /home/rory/sourcecode/cabbageaudio/csound/Top/csound.c  2972
10  0x00007ffff66a3777  csoundDestroy  /home/rory/sourcecode/cabbageaudio/csound/Top/csound.c  1244
11  0x0000000000538a9a  Csound::~Csound  /usr/local/include/csound/csound.hpp  866
12  0x0000000000538aca  Csound::~Csound  /usr/local/include/csound/csound.hpp  872

**Comment by jpffitch** [comment]
Having no idea wat you are doing it looks like locking/unlocking problems.  Needs a runnable example I think

**Comment by kunstmusik** [comment]
I took a quick look at osc.c.  I see two pairs of lock/unlock on the mutex,
but I'm not seeing a createMutex call.  Should there be one in the
osc_list_init() function?

**Comment by rorywalsh** [comment]
I was afraid you'd ask that ;) I'll see if we can prepare a simple example.

**Comment by jpffitch** [comment]
line 525 crears

**Comment by kunstmusik** [comment]
Thanks @jpffitch, I see it now.  The code was a little confusing to me the
first time.  One thing I'm wondering is if the OSC_listdeinit is running
after OSC_deinit. It might be the case that OSC_deinit is destroying the
mutex, then OSC_listdeinit tries to lock on the dead mutex.

**Comment by Oeyvind** [comment]
I'm at AES now, so will probably not manager to make a simple example right
away.  Perhaps notable that the problematic csd attempts to open a port for
listening, then if it fails increments the port number and tries again
until an available port is found.  Should not matter, as it will only open
one port. Still thought it worth mentioning just in case.

29. sep. 2016 10.32 skrev "Steven Yi" <notifications@github.com>:

**Comment by kunstmusik** [comment]
Sidenote: I just commited a fix to Engine/csound_data_structures.c that dealt with freeing up values from hash tables.  This was causing a crash when OSC_reset() was called when destroying a global variable here on Windows.  (The bug did not exhibit when I took a look on OSX.) I don't know if the bug that was fixed relates to the issue here, but thought it worth mentioning.  

@rorywalsh Could you compile the latest from the develop branch?  It'd be nice to verify whether the fix I put in is related or not.

**Comment by rorywalsh** [comment]
Thanks Steven. I will try it out and let you know.

**Comment by rorywalsh** [comment]
I'm afraid the problem persists, thanks for trying however. Here's the backtrace, although I think it's the same as the previous one. 

0  0x00007ffff6c58d84  __GI___pthread_mutex_lock  /build/glibc-GKVZIf/glibc-2.23/nptl/../nptl/pthread_mutex_lock.c  67
1  0x00007ffff66b1383  csoundLockMutex  /home/rory/sourcecode/cabbageaudio/csound/Top/threads.c  496
2  0x00007fffe1850c69  OSC_listdeinit    
3  0x00007ffff66a97a8  csoundDeinitialiseOpcodes  /home/rory/sourcecode/cabbageaudio/csound/Top/csound.c  3762
4  0x00007ffff652505e  deact  /home/rory/sourcecode/cabbageaudio/csound/Engine/insert.c  727
5  0x00007ffff6525640  xturnoff  /home/rory/sourcecode/cabbageaudio/csound/Engine/insert.c  829
6  0x00007ffff6525695  xturnoff_now  /home/rory/sourcecode/cabbageaudio/csound/Engine/insert.c  842
7  0x00007ffff6531535  deactivate_all_notes  /home/rory/sourcecode/cabbageaudio/csound/Engine/musmon.c  378
8  0x00007ffff6531723  csoundCleanup  /home/rory/sourcecode/cabbageaudio/csound/Engine/musmon.c  432
9  0x00007ffff66a7bc1  reset  /home/rory/sourcecode/cabbageaudio/csound/Top/csound.c  2972
10  0x00007ffff66a37b1  csoundDestroy  /home/rory/sourcecode/cabbageaudio/csound/Top/csound.c  1244
11  0x0000000000538a9a  Csound::~Csound  /usr/local/include/csound/csound.hpp  866
12  0x0000000000538aca  Csound::~Csound  /usr/local/include/csound/csound.hpp  872

**Comment by vlazzarini** [comment]
I think what would be helpful, if at all possible is a minimal CSD + host code demonstrating the segfault. Then we would be able to chase it with more confidence.

**Comment by Oeyvind** [comment]
Ok. Here is a simplified example. It segfaults on exit.
Do note, I have OSCinit inside and instrument, running this instrument
before anything else.
If I put OSCinit in the global instr, all works ok.

The reason I put OSCinit in an instrument was that I need my instrument to
find an available port, trying repeatedly, changing the port number if the
port is unavailable, then connecting when an available port number is
found. This is not shown in the simplified example, so for this exact code
it would naturally be better to just put OSCinit in the globals.

2016-10-01 3:15 GMT-07:00 vlazzarini <notifications@github.com>:

**Comment by jpffitch** [comment]
Where is the simplified example?  I do not see it.

**Comment by Oeyvind** [comment]
It was attached, maybe some filter stripped it off somewhere along the way.
Here is the csds text

<CsoundSynthesizer>
<CsOptions>
-n -d
</CsOptions>
<CsInstruments>

        sr = 48000
ksmps = 64
nchnls = 2
0dbfs = 1

        instr 3
gihandle OSCinit 7001
        endin

        instr 6
        krms init 0
nxt_val:
        kk1 OSClisten gihandle, "analysis", "f", krms
        if (kk1 == 0) goto ex_val
    kgoto nxt_val
ex_val:
        endin

</CsInstruments>
<CsScore>
i3 0 600
i6 1 600
e
</CsScore>
</CsoundSynthesizer>


2016-10-03 13:19 GMT-07:00 John ffitch <notifications@github.com>:

**Comment by jpffitch** [comment]
This is a user error.  Instr3 finishes at time 600 but instr 6 is running at 
601 when OSC as finished.  Setting i3 to run for 602 secs and no crash

I suppose we ought to protect against it somehow, .... maybe I am wrong

==John

**Comment by vlazzarini** [comment]
I agree, Csound should not crash but issue an error of some description.

**Comment by Oeyvind** [comment]
But does this explain why Csound crashes when performance is terminated
before the score ends? In that case,  which is the real case I tried to
simplify, the instruments both run till the end.

4. okt. 2016 05.33 skrev "John ffitch" <notifications@github.com>:

**Comment by jpffitch** [comment]
Hard to debug hyperthetical programs!

**Comment by jpffitch** [comment]
Just committed a patc that _seems_ to fix tis

**Comment by Oeyvind** [comment]
Yes, it seems to work fine here, both on the simplified example, and on the
real example (more complex, with several listeners dynamically allocated).
Thank you so much!

2016-10-04 7:00 GMT-07:00 John ffitch <notifications@github.com>:

**Comment by rorywalsh** [comment]
Thanks guys. This can be closed. Nice work ;)

**Comment by Oeyvind** [comment]
Just need to ask,
it seems that other of my instruments now may take longer to load and/or
exit.
Is there any reason the comitted change could influence startup/shutdown
time for e.g. OSCsend?

I have an instrument that uses 2 OSC sends, and if I load 3 of this
instrument, it takes much more than 3 times as long to shut down.
I should of course make a simplified example, just thought I'd ask first if
it makes sense that the recent changes could inflict such increase in
shutdown time?



2016-10-04 7:00 GMT-07:00 John ffitch <notifications@github.com>:
